### ISSUE TITLE
Language server is very slow on large document count

### ISSUE THREAD

**Issue by albnnc** [denoland/deno#21296]
<details>
<summary>Repo Description</summary>

`scc` output:

![image](https://github.com/denoland/deno/assets/20284259/42612c56-2f4c-4109-82aa-4ed896d31dc9)

JSON and XML files are generated and relate to stub data mostly, but TypeScript code is hand-written and we were really excited when document preload feature has been released. I mean, we really do want to have auto-imports and other LSP suggestions over ~1.6k TypeScript files.
</details>

Deno config has nothing special, it just ignores some directories with generated assets, none of which are included in any language server status report.

The following VSCode settings

```
"deno.documentPreloadLimit": 10000,
"deno.maxTsServerMemory": 10000,
```

lead to language server being really slow. Here is the status report:

<details>
<summary>LS status report (~4k files)</summary>

![telegram-cloud-photo-size-2-5400184504817144349-x](https://github.com/denoland/deno/assets/20284259/ae1ca33a-e7be-4650-b818-44732e7b489d)
</details>

It says that `code_action` measurement equals to ~600ms, but sometimes `Cmd + .` press freezes language server for several seconds. That being said, writing code by hand is also hard, since we have to wait seconds for VSCode to update document errors. I mean, LSP is not slow for just some sort of "document lookup", it's just _slow in general_. Moreover, sometimes code actions work correctly only after several retries, each of which wastes several seconds, but it might be a different problem, I guess.

However, if we disable document preload feature, document count drops to roughly 1k and the language server starts to respond almost instantly.

<details>
<summary>LS status report (~1k files)</summary>

![telegram-cloud-photo-size-2-5400184504817144355-x](https://github.com/denoland/deno/assets/20284259/d0181ccd-b6ce-4ec5-a828-922a93f2dcc3)
</details>

Obviously, the documents that not preloaded are unable to be imported via `Cmd + .` or similar actions. Sadly, if we open more files, more files are getting collected by language server and the problem shows again.

To me, the main problem is that small difference in file count results in much larger difference in language server responsiveness. Again, subjectively, metrics in LS status report doesn't always show the actual VSCode timings.

- Deno version is 1.38.2.
- This issue has been trailing us for a long time, not only the last version of Deno is affected.
- We're experiencing this problem on ARM Macs, Windows and Linux PCs.
- All target hosts have enough RAM, but language server doesn't use much of it actually (only a few gigs in case of ~4k documents).
- Language server restart doesn't help if document preload is set, it starts to lag in the very beginning.
- I don't know of any large file ending up being loaded by language server.

**Comment by devidw** [comment]
They claim to have made it a lot faster with 1.43 https://deno.com/blog/optimizing-our-lsp

I'm on

```
deno 2.1.4+c6fa628 (canary, release, aarch64-apple-darwin)
v8 13.0.245.12-rusty
typescript 5.6.2
```

And for a slightly bigger monorepo with deno workspace it's still very slow for me. Currently I don't benefit from intellisense at all.

**Comment by eikooc** [comment]
I am still experiencing a slow LSP in a project. I reckon it has something to do with having a lot of inferred types based on Zod schemas.

## Performance (last 3 000 entries)

|Name|Count|Duration|
|---|---|---|
|lsp.code_action|65|1374.656ms|
|lsp.code_lens|51|0.019ms|
|lsp.completion|6|156.66ms|
|lsp.completion_resolve|4|35.883ms|
|lsp.did_change|222|4.31ms|
|lsp.did_open|1|5.995ms|
|lsp.did_save|3|0.046ms|
|lsp.document_highlight|22|244.843ms|
|lsp.document_symbol|75|1602.441ms|
|lsp.folding_range|12|437.464ms|
|lsp.formatting|3|0.939ms|
|lsp.get_navigation_tree|75|1602.344ms|
|lsp.goto_definition|3|31.102ms|
|lsp.hover|18|5.296ms|
|lsp.semantic_tokens_full|24|6850.368ms|
|lsp.semantic_tokens_range|2|4317.91ms|
|lsp.signature_help|1|71.09ms|
|lsp.testing_update|224|5.177ms|
|lsp.update_diagnostics_deps|69|2.472ms|
|lsp.update_diagnostics_lint|59|34.026ms|
|lsp.update_diagnostics_ts|15|10087.818ms|
|tsc.host.$getDiagnostics|36|4856.143ms|
|tsc.host.getApplicableRefactors|125|16.201ms|
|tsc.host.getCodeFixesAtPosition|4|1.376ms|
|tsc.host.getCompletionEntryDetails|23|4.866ms|
|tsc.host.getCompletionsAtPosition|10|60.472ms|
|tsc.host.getDefinitionAndBoundSpan|5|17.948ms|
|tsc.host.getDocumentHighlights|47|8.716ms|
|tsc.host.getEncodedSemanticClassifications|22|2021.145ms|
|tsc.host.getNavigationTree|99|2.058ms|
|tsc.host.getOutliningSpans|14|1.727ms|
|tsc.host.getQuickInfoAtPosition|24|9.444ms|
|tsc.host.getSignatureHelpItems|1|3.609ms|
|tsc.op.op_is_node_file|1|0.014ms|
|tsc.op.op_load|148|0.016ms|
|tsc.op.op_resolve|1233|0.035ms|
|tsc.op.op_script_names|1|2.337ms|
|tsc.op.op_script_version|33|0.011ms|
|tsc.request.getApplicableRefactors|65|1238.096ms|
|tsc.request.getCodeFixesAtPosition|3|2943.714ms|
|tsc.request.getCompletionEntryDetails|4|35.311ms|
|tsc.request.getCompletionsAtPosition|6|129.72ms|
|tsc.request.getDefinitionAndBoundSpan|3|30.719ms|
|tsc.request.getDocumentHighlights|22|244.357ms|
|tsc.request.getEncodedSemanticClassifications|22|7865.168ms|
|tsc.request.getNavigationTree|64|1877.194ms|
|tsc.request.getOutliningSpans|12|436.355ms|
|tsc.request.getQuickInfoAtPosition|18|4.766ms|
|tsc.request.getSignatureHelpItems|1|70.909ms|

## Performance (total)

|Name|Count|Duration|
|---|---|---|
|lsp.code_action|1521|725267.849ms|
|lsp.code_lens|451|12.248ms|
|lsp.completion|43|38100.387ms|
|lsp.completion_resolve|28|3163.457ms|
|lsp.did_change|670|2832.751ms|
|lsp.did_change_watched_files|14|74409.175ms|
|lsp.did_close|50|64.576ms|
|lsp.did_open|63|230.682ms|
|lsp.did_save|32|1.590ms|
|lsp.document_highlight|784|115021.495ms|
|lsp.document_symbol|567|713633.591ms|
|lsp.folding_range|154|173577.898ms|
|lsp.formatting|23|70.479ms|
|lsp.get_navigation_tree|567|713555.240ms|
|lsp.goto_definition|275|59789.108ms|
|lsp.hover|660|3795.211ms|
|lsp.initialize|1|97.684ms|
|lsp.semantic_tokens_full|190|626357.753ms|
|lsp.semantic_tokens_range|82|192626.565ms|
|lsp.signature_help|43|18013.162ms|
|lsp.testing_update|749|3704.656ms|
|lsp.update_cache|12|0.037ms|
|lsp.update_diagnostics_deps|749|1015.773ms|
|lsp.update_diagnostics_lint|749|7629.186ms|
|lsp.update_diagnostics_ts|240|1032587.037ms|
|lsp.update_global_cache|1|560.746ms|
|lsp.virtual_text_document|13|0.291ms|
|tsc.host.$getAssets|1|6.841ms|
|tsc.host.$getDiagnostics|240|922933.319ms|
|tsc.host.$getSupportedCodeFixes|1|0.651ms|
|tsc.host.cleanupSemanticCache|22|125.577ms|
|tsc.host.getApplicableRefactors|1513|19625.479ms|
|tsc.host.getCodeFixesAtPosition|59|11543.996ms|
|tsc.host.getCompletionEntryDetails|28|3097.198ms|
|tsc.host.getCompletionsAtPosition|43|28178.200ms|
|tsc.host.getDefinitionAndBoundSpan|275|15316.740ms|
|tsc.host.getDocumentHighlights|784|19822.335ms|
|tsc.host.getEncodedSemanticClassifications|188|332928.401ms|
|tsc.host.getNavigationTree|452|880.019ms|
|tsc.host.getOutliningSpans|154|455.215ms|
|tsc.host.getQuickInfoAtPosition|659|11744.938ms|
|tsc.host.getSignatureHelpItems|43|23817.182ms|
|tsc.op.op_is_node_file|2409|4.935ms|
|tsc.op.op_load|6517|123.220ms|
|tsc.op.op_resolve|71603|4653.641ms|
|tsc.op.op_script_names|67|144.653ms|
|tsc.op.op_script_version|194|3.142ms|
|tsc.request.$getAssets|1|8.996ms|
|tsc.request.$getSupportedCodeFixes|1|82.845ms|
|tsc.request.cleanupSemanticCache|22|71434.479ms|
|tsc.request.getApplicableRefactors|1513|700584.430ms|
|tsc.request.getCodeFixesAtPosition|59|28989.076ms|
|tsc.request.getCompletionEntryDetails|28|3160.077ms|
|tsc.request.getCompletionsAtPosition|43|37840.576ms|
|tsc.request.getDefinitionAndBoundSpan|275|59713.346ms|
|tsc.request.getDocumentHighlights|784|114764.234ms|
|tsc.request.getEncodedSemanticClassifications|188|818894.622ms|
|tsc.request.getNavigationTree|452|713380.030ms|
|tsc.request.getOutliningSpans|154|173512.119ms|
|tsc.request.getQuickInfoAtPosition|659|3565.155ms|
|tsc.request.getSignatureHelpItems|43|18008.998ms|

**Comment by haydos404** [comment]
> I am still experiencing a slow LSP in a project. I reckon it has something to do with having a lot of inferred types based on Zod schemas.

I had something very identical to this in a large Deno workspace recently. Rolling back deno versions didn’t seem to help, switching from using zod, to valibot, to classes with class-validator and statically typing all of my inferred types didn’t appear to help either. I was getting completions around 45 seconds. 

I did have a svelte project in the mono-repo (which if I remember right uses ts-node under the hood for the language server) so when I purged during testing that it seemed to bump my performance a small amount. 

But unfortunately switching back to nodejs solved the poor IDE performance problem for me.

**Comment by nayeemrmn** [comment]
This is for a relatively old version of Deno. We've made a lot of improvements since then relating to request cancellation, caching and lazy starts. Please open a new issue if it's still a problem, thanks
